src/main/java/org/traccar/protocol/LaipacProtocolDecoder.java

```diff
@@ -81,81 +100,166 @@ private String decodeAlarm(String event) {
                 return Position.ALARM_GEOFENCE_EXIT;
             case "6":
                 return Position.ALARM_OVERSPEED;
+            case "5":
+                return Position.ALARM_POWER_CUT;
             case "3":
                 return Position.ALARM_SOS;
             default:
                 return null;
         }
     }
 
-    @Override
-    protected Object decode(
-            Channel channel, SocketAddress remoteAddress, Object msg) throws Exception {
+    private String decodeEvent(String event, Position position) {
 
-        String sentence = (String) msg;
+        if (event.length() == 1) {
+            char inputStatus = event.charAt(0);
+            if (inputStatus >= 'A' && inputStatus <= 'D') {
+                int inputStatusInt = inputStatus - 'A';
+                position.set(Position.PREFIX_IN + 1, inputStatusInt & 1);
+                position.set(Position.PREFIX_IN + 2, inputStatusInt & 2);
+                position.set(Position.KEY_IGNITION, ((inputStatusInt & 1) != 0) ? true : false);
+                return null;
+            }
+        }
 
-        if (sentence.startsWith("$ECHK") && channel != null) {
-            channel.writeAndFlush(new NetworkMessage(sentence + "\r\n", remoteAddress)); // heartbeat
-            return null;
+        return event;
+    }
+
+    private String getDevicePassword(DeviceSession deviceSession) {
+
+        String devicePassword = defaultDevicePassword;
+
+        Device device = Context.getIdentityManager().getById(deviceSession.getDeviceId());
+        if (device != null) {
+            String password = device.getString(Command.KEY_DEVICE_PASSWORD);
+            if (password != null) {
+                devicePassword = password;
+            }
         }
 
-        Parser parser = new Parser(PATTERN, sentence);
-        if (!parser.matches()) {
-            return null;
+        return devicePassword;
+    }
+
+    private Object handleEchk(
+            String sentence, int checksum, Parser parser, Channel channel, SocketAddress remoteAddress) {
+
+        if (channel != null) {
+            channel.writeAndFlush(new NetworkMessage(sentence + "\r\n", remoteAddress));
         }
 
-        DeviceSession deviceSession = getDeviceSession(channel, remoteAddress, parser.next());
+        DeviceSession deviceSession =
+            getDeviceSession(channel, remoteAddress, parser.next());
+        if (deviceSession != null) {
+            ConnectionManager cm = Context.getConnectionManager();
```

Comment:
Don't do this.

Reply:
The thing is that you can configure these devices to send just a heartbeat between reports. For example, you can set the device to send heartbeats every 5 minutes, and reports every 15 minutes. I thought this should be enough to state the device as on-line on every heartbeat, or on anything it sends to the server. What do you think about this? If you don't agree with me, no problem, I'll remove that code. Thanks.

Reply:
This is already handled for you. You don't need to do anything like this in the decoder.

Reply:
Ok. Removing code. Thanks.